crates/dekaf/src/topology.rs

```diff
@@ -323,40 +469,239 @@ impl Collection {
+) -> anyhow::Result<(
+    flow_client::Client,
+    AccessTokenClaims,
+    String,
+    String,
+    proto_flow::flow::MaterializationSpec,
+)> {
+    let request_token = flow_client::client::build_task_authorization_request_token(
+        shard_template_id,
+        data_plane_fqdn,
+        data_plane_signer,
+        proto_flow::capability::AUTHORIZE,
+        Default::default(),
+    )?;
+    let models::authorizations::DekafAuthResponse {
+        token,
+        ops_logs_journal,
+        ops_stats_journal,
+        task_spec,
+        retry_millis: _,
+    } = loop {
+        let response: models::authorizations::DekafAuthResponse = client
+            .agent_unary(
+                "/authorize/dekaf",
+                &models::authorizations::TaskAuthorizationRequest {
+                    token: request_token.clone(),
+                },
+            )
+            .await?;
+        if response.retry_millis != 0 {
+            tracing::warn!(
+                secs = response.retry_millis as f64 / 1000.0,
+                "authorization service tentatively rejected our request, but will retry before failing"
+            );
+            () = tokio::time::sleep(std::time::Duration::from_millis(response.retry_millis)).await;
+            continue;
+        }
+        break response;
+    };
+    let claims = flow_client::parse_jwt_claims(token.as_str())?;
+
+    Ok((
+        client.with_user_access_token(Some(token)),
+        claims,
+        ops_logs_journal,
+        ops_stats_journal,
+        serde_json::from_str(
+            task_spec
+                .ok_or(anyhow::anyhow!(
+                    "task_spec is only None when we need to retry the auth request"
+                ))?
+                .get(),
+        )?,
+    ))
+}
+
+pub async fn extract_dekaf_config(
+    spec: &proto_flow::flow::MaterializationSpec,
+) -> anyhow::Result<DekafConfig> {
+    let config = serde_json::from_str::<models::DekafConfig>(&spec.config_json)?;
+
+    let decrypted_endpoint_config =
+        unseal::decrypt_sops(&RawValue::from_str(&config.config.to_string())?).await?;
+
+    let dekaf_config = serde_json::from_str::<DekafConfig>(&decrypted_endpoint_config.to_string())?;
+    Ok(dekaf_config)
+}
+
+/// Nests the provided shape under a JSON pointer path by creating the necessary object hierarchy.
+/// For example, given pointer "/a/b/c" and a field shape, creates an object structure:
+/// { "a": { "b": { "c": field_shape } } }
+fn build_shape_at_pointer(ptr: &doc::Pointer, shape: &doc::Shape) -> doc::Shape {
+    // Return the original shape if pointer is empty
+    if ptr.0.is_empty() {
+        return shape.clone();
+    }
+
+    tracing::debug!(?ptr, "Building shape");
+    let mut current_shape = doc::Shape::nothing();
+    let mut current = &mut current_shape;
+
+    // For each component in the pointer path except the last one,
+    // create the object structure
+    for token in ptr.iter().take(ptr.0.len() - 1) {
```

Comment:
This is hard to follow even just a few days after writing it, refactor to be a fold over `ptr.iter()` with `doc::Shape::nothing()` as the initial state